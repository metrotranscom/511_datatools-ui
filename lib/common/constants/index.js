// @flow

const SECURE: string = 'secure/'
export const API_PREFIX: string = `/api/manager/`
export const SECURE_API_PREFIX: string = `${API_PREFIX}${SECURE}`
export const GTFS_GRAPHQL_PREFIX: string = `${SECURE_API_PREFIX}gtfs/graphql`
export const EDITOR_PREFIX: string = `/api/editor/`
export const SECURE_EDITOR_PREFIX: string = `${EDITOR_PREFIX}${SECURE}`
export const DEFAULT_DESCRIPTION = 'A command center for managing, editing, validating, and deploying GTFS.'
export const DEFAULT_LOGO = 'https://d2tyb7byn1fef9.cloudfront.net/ibi_group_black-512x512.png'
export const DEFAULT_LOGO_SMALL = 'https://d2tyb7byn1fef9.cloudfront.net/ibi_group-128x128.png'
export const DEFAULT_TITLE = 'Data Tools'

export const RETRIEVAL_METHODS = Object.freeze({
  MANUALLY_UPLOADED: 'MANUALLY_UPLOADED',
  FETCHED_AUTOMATICALLY: 'FETCHED_AUTOMATICALLY',
  PRODUCED_IN_HOUSE: 'PRODUCED_IN_HOUSE',
  SERVICE_PERIOD_MERGE: 'SERVICE_PERIOD_MERGE',
  REGIONAL_MERGE: 'REGIONAL_MERGE',
  VERSION_CLONE: 'VERSION_CLONE'
})

export const FEED_TRANSFORMATION_TYPES = Object.freeze({
  DELETE_RECORDS: 'DeleteRecordsTransformation',
  REPLACE_FILE_FROM_VERSION: 'ReplaceFileFromVersionTransformation',
  REPLACE_FILE_FROM_STRING: 'ReplaceFileFromStringTransformation'
})
