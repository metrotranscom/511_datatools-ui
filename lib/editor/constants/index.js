// @flow

export const ENTITY = {
  // TODO: use these constants for component names
  STOP: 'stop',
  ROUTE: 'route',
  TRIP_PATTERN: 'trippattern',

  // For constructing new entities (before saving to server)
  NEW_ID: -2
}

export const POINT_TYPE = Object.freeze({
  DEFAULT: 0,
  ANCHOR: 1,
  STOP: 2
})

export const ARROW_MAGENTA = '#d50b65'
