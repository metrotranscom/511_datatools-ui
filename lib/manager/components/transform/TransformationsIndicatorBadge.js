// @flow

import Icon from '@conveyal/woonerf/components/icon'
import React, {Component} from 'react'
import {OverlayTrigger, Tooltip} from 'react-bootstrap'

import type { FeedVersion } from '../../../types'

type Props = {
  version: FeedVersion
}

type State = {
  expanded: boolean
}

/**
 * Renders a badge with tooltip showing a summary of the transformations applied
 * to the feed version during processing by the server backend.
 */
export default class TransformationsIndicatorBadge extends Component<Props, State> {
  state = {expanded: false}

  _renderTooltipContent = () => {
    const {version} = this.props
    const {feedTransformResult} = version
    if (!feedTransformResult) return null
    const transformationCount = feedTransformResult.tableTransformResults.length

    return (
      <div style={{paddingLeft: '2px'}}>
        Feed has {transformationCount} transformation(s):
        <ul>
          {feedTransformResult.tableTransformResults.map((item, i) => {
            return (
              <li key={i}>
                {item.tableName}{' '}
                {item.transformType.toLowerCase().replace('_', ' ')}
              </li>
            )
          })}
        </ul>
      </div>
    )
  }

  render () {
    const { version } = this.props
    if (!version.feedTransformResult) return null
    return (
      <OverlayTrigger
        overlay={
          <Tooltip id='transformations-indicator'>
            {this._renderTooltipContent()}
          </Tooltip>
        }>
        <Icon
          type='wrench'
        />
      </OverlayTrigger>
    )
  }
}
