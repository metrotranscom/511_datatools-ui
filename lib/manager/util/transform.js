// @flow
import {getComponentMessages} from '../../common/util/config'

import type {FeedTransformation, FeedVersion} from '../../types'

/**
 * Returns human readable name or description (if transformation object provided)
 * for transformation type based on messages defined in
 * english.yml#components.FeedTransformationDescriptions.
 */
export function getTransformationName (
  type: string,
  transformation?: FeedTransformation,
  versions?: Array<FeedVersion>
) {
  // Get the base component messages.
  const messages = getComponentMessages('FeedTransformationDescriptions')
  // If no transformation is provided, the name of the transformation is being
  // requested.
  if (!transformation) return messages(`${type}.name`)
  if (type !== transformation['@type']) {
    console.error(`Input feed transformation type (${type}) does not match type found in transformation object`, transformation)
  }
  // Otherwise, build the label/description up piece by piece.
  const {csvData, table, sourceVersionId} = transformation
  let label = messages(`${type}.label`)
  // Note: Additional placeholder values can be added to this list and the below
  // switch statement.
  const placeholders = ['filePlaceholder', 'tablePlaceholder', 'versionPlaceholder']
  placeholders.forEach(placeholderId => {
    // Replacement text defaults to the placeholder value.
    let replacementText = messages(`general.${placeholderId}`)
    // But if the respective field has data defined, override the placeholder
    // value below.
    switch (placeholderId) {
      case 'filePlaceholder':
        if (csvData) replacementText = messages(`general.fileDefined`)
        break
      case 'tablePlaceholder':
        if (table) replacementText = `${table} ${messages(`general.table`)}`
        break
      case 'versionPlaceholder':
        const version = versions && versions.find(v => v.id === sourceVersionId)
        if (version) replacementText = `${messages(`general.version`)} ${version.version}`
        break
      default:
        console.warn(`No case defined for placeholder field: ${placeholderId}`)
        break
    }
    label = label.replace(`%${placeholderId}%`, replacementText)
  })
  return label
}
