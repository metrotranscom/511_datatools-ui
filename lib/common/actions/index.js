// @flow

import fetch from 'isomorphic-fetch'

import {GTFS_GRAPHQL_PREFIX} from '../../common/constants'
import { setErrorMessage } from '../../manager/actions/status'

import type {dispatchFn, getStateFn} from '../../types/reducers'

export function createVoidPayloadAction (type: string) {
  return () => ({ type })
}

export function secureFetch (url: string, method: string = 'get', payload?: any, raw: boolean = false, isJSON: boolean = true, actionOnFail?: string): any {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    function consoleError (message) {
      console.error(`Error making ${method} request to ${url}: `, message)
    }

    // if running in a test environment, fetch will complain when using relative
    // urls, so prefix all urls with http://localhost:4000.
    if (process.env.NODE_ENV === 'test') {
      url = `http://localhost:4000${url}`
    }

    let {token} = getState().user
    // FIXME What if authentication is turned off.
    if (!token) {
      console.warn('Cannot fetch without user token.')
      token = 'no_auth'
    }
    const headers: {[string]: string} = {
      'Authorization': `Bearer ${token}`,
      'Accept': 'application/json'
    }
    if (isJSON) {
      headers['Content-Type'] = 'application/json'
    }
    const body = payload && isJSON
      ? JSON.stringify(payload)
      : payload
    return fetch(url, {method, headers, body})
      // Catch basic error during fetch
      .catch(err => {
        const message = `Error making request: (${err})`
        consoleError(message)
        return dispatch(setErrorMessage({message}))
      })
      .then(res => {
        // if raw response is requested
        if (raw) return res
        let action, message
        // check for errors
        const {status} = res
        if (status >= 500) {
          action = 'RELOAD'
          message = `Network error!\n\n(${method} request on ${url})`
          consoleError(message)
          dispatch(setErrorMessage({message, action}))
          return null
        } else if (status >= 400) {
          action = status === 401
            ? 'LOG_IN'
            : actionOnFail
          res.json()
            .then(json => {
              const {detail, message} = json
              const unknown = `Unknown (${status}) error occurred while making request`
              consoleError(message || JSON.stringify(json) || unknown)
              dispatch(setErrorMessage({
                message: message || JSON.stringify(json) || unknown,
                action,
                detail
              }))
            })
          return null
        } else {
          return res
        }
      })
  }
}

function graphQLErrorsToString (errors: Array<{locations: any, message: string}>): Array<string> {
  return errors.map(e => {
    const locations = e.locations.map(JSON.stringify).join(', ')
    return `- ${e.message} (${locations})\n`
  })
}

export function fetchGraphQL ({
  errorMessage,
  query,
  variables
}: {
  errorMessage?: string,
  query: string,
  variables?: {[key: string]: string | number | Array<string>}
}): any {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const body = {
      query,
      variables
    }
    return dispatch(secureFetch(GTFS_GRAPHQL_PREFIX, 'post', body))
      .then(res => res.json())
      .then(json => {
        if (json.errors && json.errors.length) {
          dispatch(setErrorMessage({
            message: errorMessage || 'Error fetching GTFS entities via GraphQL',
            // action,
            detail: graphQLErrorsToString(json.errors)
          }))
        } else {
          return json.data
        }
      })
      .catch(err => console.log(err))
  }
}
