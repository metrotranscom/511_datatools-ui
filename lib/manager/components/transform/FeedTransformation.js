// @flow

import Icon from '@conveyal/woonerf/components/icon'
import React, {Component} from 'react'
import {Button} from 'react-bootstrap'
import Select from 'react-select'

import {getGtfsSpec, getGtfsPlusSpec, isModuleEnabled} from '../../../common/util/config'
import {getTransformationName} from '../../util/transform'
import VersionSelectorDropdown from '../version/VersionSelectorDropdown'

import type {
  Feed,
  FeedTransformation as FeedTransformationType,
  ReactSelectOption
} from '../../../types'

type Props = {
  feedSource: Feed,
  index: number,
  onChange: (any, number) => void,
  onRemove: (number) => void,
  transformation: FeedTransformationType
}

/**
 * Component that renders fields for a single feed transformation
 * (e.g., ReplaceFileFromStringTransformation).
 */
export default class FeedTransformation extends Component<Props, {csvData: ?string}> {
  state = { csvData: null }

  _getFieldsForType = (type: string) => {
    const {feedSource, transformation} = this.props
    // Get selected version (if applicable).
    const version = feedSource.feedVersions &&
      feedSource.feedVersions.find(v => v.id === transformation.sourceVersionId)
    const fields = []
    let index = 0
    switch (type) {
      case 'ReplaceFileFromVersionTransformation':
        fields.push(
          <VersionSelectorDropdown
            key={index++}
            dropdownProps={{
              id: 'merge-versions-dropdown',
              onSelect: this._onSelectVersion
            }}
            title={version
              ? `From source version ${version.version}`
              : 'Select a source version'
            }
            version={version}
            versions={feedSource.feedVersions}
          />
        )
        break
      case 'ReplaceFileFromStringTransformation':
        const inputIsUnchanged = this.state.csvData === null
        const csvData = inputIsUnchanged
          ? transformation.csvData
          : this.state.csvData
        const textValue = csvData || ''
        const numLines = !textValue ? 0 : textValue.split(/\r*\n/).length
        fields.push(
          <div>
            <label htmlFor='csvData'>
              Add the CSV data to add to/replace in the incoming GTFS:
              <textarea
                id='csvData'
                onChange={this._onChangeCsvData}
                style={{
                  fontFamily: 'monospace',
                  fontSize: 'x-small',
                  height: '80px',
                  overflow: 'auto',
                  width: '400px',
                  whiteSpace: 'pre'
                }}
                placeholder={
                  `stop_id,stop_code,stop_name,stop_lat,stop_lon
  1234567,188390987,Broad Ave,33.98768,-87.72686`
                }
                value={textValue} />
            </label>
            <div style={{marginBottom: '10px'}}>
              <Button
                bsSize='xsmall'
                disabled={inputIsUnchanged}
                style={{marginRight: '5px'}}
                onClick={this._onSaveCsvData}>
                Save CSV
              </Button>
              <small>{numLines} lines</small>
            </div>
          </div>
        )
        break
      default:
        break
    }
    return fields
  }

  _getValidationIssues = () => {
    const {feedSource, transformation} = this.props
    // Get selected version (if applicable).
    const version = feedSource.feedVersions &&
      feedSource.feedVersions.find(v => v.id === transformation.sourceVersionId)
    const issues = []
    if (!transformation.table) {
      issues.push('Table must be defined')
    }
    switch (transformation['@type']) {
      case 'ReplaceFileFromVersionTransformation':
        // Only trigger validation issue if versions have been loaded.
        if (!version && feedSource.feedVersions) {
          issues.push('Version must be defined')
        }
        break
      case 'ReplaceFileFromStringTransformation':
        if (!transformation.csvData) {
          issues.push('CSV Data must be defined')
        }
        // TODO: Further validate CSV data based on parsing fields, checking for
        //   missing columns, etc.
        break
      default:
        break
    }
    return issues
  }

  _onChangeCsvData = (evt: SyntheticInputEvent<HTMLInputElement>) => {
    this.setState({csvData: evt.target.value})
  }

  _onSaveCsvData = () => {
    const csvData = this.state.csvData || null
    this.setState({csvData: null})
    this.props.onChange({csvData}, this.props.index)
  }

  _onRemoveTransformation = () => {
    this.props.onRemove(this.props.index)
  }

  _onSelectTable = (tableOption: ReactSelectOption) => {
    this.props.onChange({table: tableOption.value}, this.props.index)
  }

  _onSelectVersion = (versionIndex: number) => {
    const {feedSource, onChange} = this.props
    if (feedSource.feedVersions) {
      const version = feedSource.feedVersions[versionIndex - 1]
      onChange({sourceVersionId: version.id}, this.props.index)
    } else {
      console.warn('Feed source does not have list of feed versions')
    }
  }

  render () {
    const {feedSource, index, transformation} = this.props
    const tables = [...getGtfsSpec()].filter(t => !t.datatools)
    if (isModuleEnabled('gtfsplus')) tables.push(...getGtfsPlusSpec())
    const transformationType = transformation['@type']
    const validationIssues = this._getValidationIssues()
    const backgroundColor = validationIssues.length === 0
      ? '#f7f7f7'
      : '#ffdf68'
    return (
      <div style={{
        backgroundColor,
        padding: '2px 10px 10px 10px',
        marginBottom: '10px',
        border: '#ccc 1px solid',
        borderRadius: '10px'
      }}>
        <h5>
          <Button
            bsSize='xsmall'
            className='pull-right'
            onClick={this._onRemoveTransformation}>
            Delete step
          </Button>
          Step {index + 1} - {getTransformationName(transformationType, transformation, feedSource.feedVersions)}
        </h5>
        <div className='feed-transformation-body'>
          <Select
            style={{margin: '10px 0px', width: '230px'}}
            clearable={false}
            placeholder='Choose the file/table to replace'
            options={tables.map(table => ({value: table.name.split('.txt')[0], label: table.name}))}
            value={transformation.table}
            onChange={this._onSelectTable} />
          {this._getFieldsForType(transformationType)}
          {validationIssues.length > 0
            ? <ul className='list-unstyled'>
              <Icon type='exclamation-triangle' /> <strong>Validation Issues</strong>
              {validationIssues.map((issue, i) => {
                return (
                  <li style={{marginLeft: '10px'}} key={`issue-${i}`}>
                    {issue}
                  </li>
                )
              })}
            </ul>
            : null
          }
        </div>
      </div>
    )
  }
}
