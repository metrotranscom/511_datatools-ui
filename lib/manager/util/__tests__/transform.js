// @flow
import clone from 'lodash/cloneDeep'

import {getTransformationName} from '../transform'
import {FEED_TRANSFORMATION_TYPES} from '../../../common/constants'

import type {FeedTransformation} from '../../../types'

const {REPLACE_FILE_FROM_STRING} = FEED_TRANSFORMATION_TYPES

const replaceFileTransformation: FeedTransformation = {
  '@type': REPLACE_FILE_FROM_STRING,
  active: true,
  sourceVersionId: 'abcdefg-12345.zip',
  table: null
}

const halfCompletedTransformation = clone(replaceFileTransformation)
halfCompletedTransformation.csvData = 'feed_id\n1'

const fullyCompletedTransformation = clone(halfCompletedTransformation)
fullyCompletedTransformation.table = 'feed_info'

describe('manager > util > transform >', () => {
  describe('getTransformationName >', () => {
    it('should return name for ReplaceFileFromStringTranformation', () => {
      expect(getTransformationName(REPLACE_FILE_FROM_STRING)).toMatchSnapshot()
    })

    it('should return unreplaced label for ReplaceFileFromStringTranformation', () => {
      expect(
        getTransformationName(
          REPLACE_FILE_FROM_STRING,
          replaceFileTransformation
        )
      ).toMatchSnapshot()
    })

    it('should return half-replaced label for ReplaceFileFromStringTranformation', () => {
      expect(
        getTransformationName(
          REPLACE_FILE_FROM_STRING,
          halfCompletedTransformation
        )
      ).toMatchSnapshot()
    })

    it('should return fully-replaced label for ReplaceFileFromStringTranformation', () => {
      expect(
        getTransformationName(
          REPLACE_FILE_FROM_STRING,
          fullyCompletedTransformation
        )
      ).toMatchSnapshot()
    })
  })
})
