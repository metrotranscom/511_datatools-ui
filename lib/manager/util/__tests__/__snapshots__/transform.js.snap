// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > util > transform > getTransformationName > should return fully-replaced label for ReplaceFileFromStringTranformation 1`] = `"Replace feed_info table from below text"`;

exports[`manager > util > transform > getTransformationName > should return half-replaced label for ReplaceFileFromStringTranformation 1`] = `"Replace [choose table] from below text"`;

exports[`manager > util > transform > getTransformationName > should return name for ReplaceFileFromStringTranformation 1`] = `"Replace file from string transformation"`;

exports[`manager > util > transform > getTransformationName > should return unreplaced label for ReplaceFileFromStringTranformation 1`] = `"Replace [choose table] from [choose file]"`;
